src/Server/Event/Listener/CollectionListener.php

```diff
@@ -0,0 +1,209 @@
+            if ($metadata->isCollectionValuedAssociation($association)) {
+                // Skip handling associations that aren't in the data
+                if (array_key_exists($association, $originalData)
+                    && !empty( $originalData[$association] )
+                    && ( is_array($originalData[$association])
+                         || $originalData[$association] instanceof \Traversable )
+                ) {
+                    // Ensure the collection value has an input filter
+                    if (!$inputFilter->has($association)) {
+                        /*
+                         * If we got here it means the value wasn't in the input filter and wasn't stripped out.
+                         * Treat as hostile and stop execution.
+                         */
+                        throw new InvalidArgumentException('Non-validated input detected');
+                    }
+
+
+                    foreach ($originalData[$association] as &$subEntityData) {
+                        $target          = $metadata->getAssociationTargetClass($association);
+                        $identifierNames = $metadata->getIdentifierFieldNames($target);
+                        if (empty( $identifierNames )) {
+                            continue;
+                        }
+
+                        $identifierValues = [ ];
+                        foreach ($identifierNames as $identifierName) {
+                            if (!isset( $subEntityData[$identifierName] ) || empty( $subEntityData[$identifierName] )) {
+                                continue; // Should mean we are working with a new entity to be created
+                            }
+                            $identifierValues[$identifierName] = $subEntityData[$identifierName];
+                        }
+
+                        $subEntity = false;
+                        if (count($identifierValues) === count($identifierNames)) {
+                            $subEntity = $objectManager->find($target, $identifierValues);
+                        }
+
+                        if (!$subEntity) {
+                            $subEntity = new $target;
+                        }
+
+                        $hydrator = $this->getEntityHydrator($target, $objectManager);
+                        $hydrator->hydrate($subEntityData, $subEntity);
+                        $objectManager->persist($subEntity);
+
+                        // Replace the data with the entity and let the downstream parent hydrator handle persistence
+                        $subEntityData = $subEntity;
+                    }
+                }
+            }
+        }
+
+        $event->setData($originalData);
+
+        return $originalData;
+    }
+
+    /**
+     * @param $entityClass
+     *
+     * @param $objectManager
+     *
+     * @return AbstractHydrator|DoctrineObject
+     */
+    protected function getEntityHydrator($entityClass, $objectManager)
```

Comment:
It might make sense to have @weierophinney take a look at this piece in particual which I'm using to fetch a doctrine hydrator if its in defined in config. Not sure if it makes more sense to build this out in our module or in zf-doctrine-hydration-module